### PR TITLE
MessageBrokerImpl implementation

### PR DESCRIPTION
**Pull request by Hroy02**

### CODE REVIEW

File: src/main/java/bgu/spl/mics/MessageBrokerImpl.java [hunk 1]
```diff
@@ -56,19 +63,24 @@ public void subscribeBroadcast(Class<? extends Broadcast> type, Subscriber m) {
...
 	/**
 	 * @pre none
 	 * @post all subscribers subscribed to e.getClass() queue's last element will be e.
 	 */
 	public <T> Future<T> sendEvent(Event<T> e) {
-		// TODO Auto-generated method stub
-		return null;
+		Subscriber currentSub = _eventSubscribers.get(e.getClass()).poll();
+		Future<T> f = new Future<>();
+		_eventFutures.put(e, f);
```
Comment (IdanTo1) [inline src/main/java/bgu/spl/mics/MessageBrokerImpl.java]: * I think it should go to the end of the method. 
* is this `put` thread-safe? I don't really know concurrentHashMap implementation..

File: src/main/java/bgu/spl/mics/MessageBrokerImpl.java [hunk 2]
```diff
@@ -77,28 +89,32 @@ public void sendBroadcast(Broadcast b) {
...
 
 	@Override
 	/**
 	 * @post: no queue associated with m
 	 */
 	public void unregister(Subscriber m) {
-		// TODO Auto-generated method stub
-
+		// Resolve all event's futures assigned to m with null, to avoid infinite wait for these futures
+		for(Object message : _subscriberQueues.get(m)) {
```
Comment (IdanTo1) [inline src/main/java/bgu/spl/mics/MessageBrokerImpl.java]: why not Message message?

File: src/main/java/bgu/spl/mics/MessageBrokerImpl.java [hunk 3]
```diff
@@ -77,28 +89,32 @@ public void sendBroadcast(Broadcast b) {
...
+			if(message instanceof Event) {
```
Comment (IdanTo1) [inline src/main/java/bgu/spl/mics/MessageBrokerImpl.java:107]: nice!

**Review by IdanTo1 — CHANGES_REQUESTED** [review]
looks perfect! please add **a lot of documentation** about concurrency and thread-safty

File: src/main/java/bgu/spl/mics/MessageBrokerImpl.java (hunk 1 above)
Comment (Hroy02) [reply src/main/java/bgu/spl/mics/MessageBrokerImpl.java]: - Done
- The put is thread-safe when using Concurrenthashmap, which we are

File: src/main/java/bgu/spl/mics/MessageBrokerImpl.java (hunk 2 above)
Comment (Hroy02) [reply src/main/java/bgu/spl/mics/MessageBrokerImpl.java]: Changed

File: src/main/java/bgu/spl/mics/MessageBrokerImpl.java [hunk 4]
```diff
@@ -77,9 +77,9 @@ public void sendBroadcast(Broadcast b) {
 	public <T> Future<T> sendEvent(Event<T> e) {
 		Subscriber currentSub = _eventSubscribers.get(e.getClass()).poll();
 		Future<T> f = new Future<>();
```
Comment (IdanTo1) [inline src/main/java/bgu/spl/mics/MessageBrokerImpl.java]: this as well :sweat_smile:

**Review by IdanTo1 — CHANGES_REQUESTED** [review]

**Review by IdanTo1 — CHANGES_REQUESTED** [review]
perfect! just move this new Future down :sweat_smile: 
these both sections should be logically separated

File: src/main/java/bgu/spl/mics/MessageBrokerImpl.java (hunk 4 above)
Comment (Hroy02) [reply src/main/java/bgu/spl/mics/MessageBrokerImpl.java]: Done

File: src/main/java/bgu/spl/mics/MessageBrokerImpl.java [hunk 5]
```diff
@@ -101,6 +101,7 @@ public void register(Subscriber m) {
 	 * @post: no queue associated with m
 	 */
 	public void unregister(Subscriber m) {
+		if(_subscriberQueues.get(m) == null) return;
```
Comment (IdanTo1) [inline src/main/java/bgu/spl/mics/MessageBrokerImpl.java:104]: I'm not sure it is needed, but no harm

**Review by IdanTo1 — APPROVED** [review]

**Pull request merged**